openstack/compute/v2/availabilityzones/utils.go

```diff
@@ -0,0 +1,35 @@
+package availabilityzones
+
+import (
+	"errors"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/availabilityzones"
+)
+
+// AZExists is a convienience function that checks if an availability zone exists and is
+// actually available.
+func AZExists(client *gophercloud.ServiceClient, name string) error {
```

Comment:
In this function, you get allPages for each call. This means that if you have to check multiple AZs on the caller side, calling this function repeatedly calls get expensive.

If I had to choose a function signature for this helper, I'd choose one that returns a map:

```
func AvailabilityZones(client *gophercloud.ServiceClient) ([map]availabilityzones.AvailabilityZone, error)
```

This results in the same number of API calls per function call, however you can use the data multiple times on the caller side:

```
availabilityZones, _ := utils.AvailabilityZones(client) // check the error :)
if _,ok:=availabilityZones(mightExistOrNot);ok {
    # zone exists
}

Reply:
Rather than returning a map where a key is the AZ name and the value is the AZ data, I think returning a list of available AZs is enough.
The reason is that AZ don't have much infos, see:

```
 openstack availability zone list
+-----------+-------------+
| Zone Name | Zone Status |
+-----------+-------------+
| nova      | available   |
+-----------+-------------+
```

So I propose that we create an helper that returns the AZs which are actually available:

```go
// ListAvailableAvailabilityZones is a convienience function that return a slice of
// available Availability Zones.
func ListAvailableAvailabilityZones(client *gophercloud.ServiceClient) ([]string, error) {
	var ret []string
	allPages, err := availabilityzones.List(client).AllPages()
	if err != nil {
		return ret, err
	}

	availabilityZoneInfo, err := availabilityzones.ExtractAvailabilityZones(allPages)
	if err != nil {
		return ret, err
	}

	for _, zoneInfo := range availabilityZoneInfo {
		if zoneInfo.ZoneState.Available {
			ret = append(ret, zoneInfo.ZoneName)
		}
	}
	return ret, nil
}
```